### ISSUE TITLE
Can't get the networked version of scream working

### ISSUE THREAD

**Issue by botiapa** [duncanthrax/scream#60]
When I first download scream, I assumed that the IVSHMEM version would be faster (because there wouldn't be any networking involved) and it worked perfectly, but as I later found out the normal version has noticeably less latency. When I tried to switch back though, I couldn't get it to work. Here's what I tried:

- Removing the IVSHMEM entry from regedit
- Completely reinstalling driver with restarting VM multiple times in between

The guest is a windows 10 VM running in qemu, it's using a virtio networking interface.
The host in running linux mint 19.2 and I'm using the included pulseaudio receiver.
I can reach the guest from the host and also the host from the guest via a ping command.

What happens:
The receiver starts (no output, doesn't crash) but I can't hear any output. It does open a sound output though on the host.

Let me know if you need more information about this issue.

**Comment by ivan** [comment]
If you are using multicast, Windows might be sending it over the wrong interface.

Does it work with unicast? Any firewall on the Linux side?

**Comment by botiapa** [comment]
I haven't tried it with unicast as I read somewher that it is much worse quality (Can try it later though) I disabled the firewall on both linux and windows.

**Comment by duncanthrax** [comment]
With which options do you run the receiver? (Also you need to use the networked receiver - just making sure :) ). You might want to specify the interface with -i option. 

You can also do a tcpdump on the interface pointing to the guest, like so:

```tcpdump -pni vethX port 4010```

where "vethX" is the name of the interface pointing to the guest. When sound plays, you should see a stream of UDP packets.

**Comment by TacticalTux** [comment]
I'm attempting to use the networked version as well (as IVSHMEM didn't work for me either,) and I am receiving no UDP packets.

**Comment by botiapa** [comment]
I use this command to start the receiver:
`./scream-pulse  -i virbr0`
Here's a tcp dump:
https://pastebin.com/EL3wSsbd
command : `user@comp:~$ sudo tcpdump -pni virbr0 port 4010 > Downloads/tcpdumb.txt`

If I understand correctly, there is data coming through but some reason the player doesnt play it.

**Comment by duncanthrax** [comment]
Yeah that is sound coming in all right. Next step is to check if the receiver gets and plays it.

Run the receiver in strace (as root), like:

```strace ./scream-pulse -i virbr0```

You might have to install strace.

Once you start playing sound, the output should show hectic activity. 

If that happens, and you still don't hear a thing, maybe the stream is at volume zero, or you have another local problem.

**Comment by botiapa** [comment]
Here's the output of the strace: https://pastebin.com/Bsda4jA6

I see a lot of not found errors, but I don't see any 'hectic' activity. Could it be that apparmor is blocking the utility?

**Comment by botiapa** [comment]
Something is wrong with the receiver then right?

**Comment by duncanthrax** [comment]
Yeah some firewalling (or apparmor) could be the issue. The receiver does not get the packets that show up on the wire.

**Comment by bernhard263** [comment]
Here are some tips for troubleshooting in a Windows only environment (I used Windows 10).

Download a packet sniffer, an easy to use one would be this: https://www.nirsoft.net/utils/smsniff.html
Or Microsoft Network Monitor (not as easy to use): https://www.microsoft.com/en-us/download/details.aspx?id=4865

Start the packet sniffer on both machines and play sound. First, check the machine that is broadcasting the sound. You should see an UDP broadcast stream to "239.255.77.77:4010" with lots of activity, several hundred kb/s. Then, check the machine that shall receive the sound. You should see the same. If not, temporarily disable any Software Firewall and see if that makes it work.

The sending machine needs to allow outgoing UDP traffic on port 4010, but the Windows firewall allows all outgoing traffic by default, so this should work right away.

On the receiving machine, all incoming traffic is blocked by default, so you need to create an inbound rule to allow UDP traffic on port 4010 for the playback software. The ScreamReader software that is included with Scream should automatically request such a rule when you first start it. Check your firewall settings whether such a rule exists.

To play back the stream, VLC player did not work well for me. It refused to play the multicast stream, only the single cast worked, and it was extremely choppy. The ScreamReader software worked great.

**Comment by JaneSmith** [comment]
I think I'm having the same issue. I'm running a similar setup, with a Windows 10 VM in QEMU, and I initially used IVSHMEM assuming it would be better. I've since discovered that it has horrendous latency, and am trying to switch to the networked version. I can't get it to work. There's no audio from the receiver. I've removed the UseIVSHMEM registry entry, tried reinstalling Scream on Windows, tried messing with unicast IPs and ports, I just can't get it to work.

I'm running Fedora Silverblue 31, which doesn't have AppArmor. Instead, it has SELinux, but my SELinux Troubleshooter tool doesn't detect any issues with Scream.

**Comment by martinellimarco** [comment]
A couple of years ago I remember having troubles making the network version work. The IVSHMEM version wasn't a thing then.

At the beginning I wasn't seeing packets from the VM all the time, because I was using 2 different network cards, one for the public internet and one for host-guest communication. Windows was picking one card at random at boot. The solution was to set the host-guest network card with a lower priority. Later unicast was implemented, that's another way to fix this.

The second issue I was having was the linux firewall. With the network card priority fixed I was seeing all the incoming packets but I couldn't make it work for the life of me. 
I disabled UFW and messed a lot with iptables for nothing.
In the end I discovered that on a recent update ubuntu installed and activated firewalld.
A simple ```systemctl stop firewalld``` and scream started working.

Moral of the story is that if you see packets with tcpdump but the app doesn't receive them then you should really check the firewall.

**Comment by canselcik** [comment]
I am also having the same issue. Works through ivshmem but latency is too high for comfortable gaming. The only reason I fell back to using ivshmem was actually this exact issue described here to begin with.

I can see the UDP multicast on wireshark but scream can't read it.

**Comment by botiapa** [comment]
Well, I finally switched to arch, and the problem is fixed for me. Sadly I don't know what fixed it.

**Comment by bcHelix** [comment]
Alright, I give up. I've tried to get this working for about 5 hours now.

The Windows VM is setup to use unicast 172.16.1.1:4010. I'm getting packets through the bridge just fine (virbr1). The Scream command is: `./scream -u -i virbr1 -p 4010 -o pulse` and using it with `-P` makes no difference. I can see audio being played in the Windows VM and there's a Scream output device in Pulse. I also get `Switched format to sample rate 48000, sample size 32 and 2 channels.` once some audio is playing so it seems Scream is getting _something_. I've checked the audio levels, everything is fine - just no audio. Any ideas?

**Comment by martinellimarco** [comment]
@bcHelix if you get the "switched format" message than you are receiving the audio packets. The format is in the header of each packet.

Can you try `-o alsa` instead of `-o pulse` to see if it makes any difference?

**Comment by bcHelix** [comment]
Same result with `-o alsa`. I verified everything was good in `alsamixer`.

**Comment by botiapa** [comment]
Sadly I can't comment on the issue, because I switched to arch, and it works fine for me now. It was probably something to do with a firewall or something.

**Comment by martinellimarco** [comment]
@bcHelix can you try with `-o raw > out.raw` instead of `-o pulse` and then send me the out.raw file?

**Comment by bcHelix** [comment]
`./scream -u -i virbr1 -p 4010 -o raw > ~/out.raw` produces a zero-byte file unfortunatly. I can still see tcpdump giving me (.153 is the VM):

`tcpdump: verbose output suppressed, use -v or -vv for full protocol decode`
`listening on virbr1, link-type EN10MB (Ethernet), capture size 262144 bytes`
`09:30:43.574420 IP 172.16.1.153.4010 > 172.16.1.1.4010: UDP, length 1157`
`09:30:43.575264 IP 172.16.1.153.4010 > 172.16.1.1.4010: UDP, length 1157`
`09:30:43.575349 IP 172.16.1.153.4010 > 172.16.1.1.4010: UDP, length 1157`
`09:30:43.575408 IP 172.16.1.153.4010 > 172.16.1.1.4010: UDP, length 1157`
`09:30:43.575454 IP 172.16.1.153.4010 > 172.16.1.1.4010: UDP, length 1157`
`09:30:43.575501 IP 172.16.1.153.4010 > 172.16.1.1.4010: UDP, length 1157`
`09:30:43.575554 IP 172.16.1.153.4010 > 172.16.1.1.4010: UDP, length 1157`
`09:30:43.575763 IP 172.16.1.153.4010 > 172.16.1.1.4010: UDP, length 1157`
`...`

I had a hunch Apparmor was doing something here, but I checked the log for any apparmor DENIED messages and didn't see any.

**Comment by martinellimarco** [comment]
@bcHelix just to be clear, you see the "Switched format" message when you try with raw?

Do you have any firewall enabled? I don't know what's your distro, try `sudo systemctl status firewalld.service`, `sudo systemctl status firewall.service` or `sudo ufw status`. If one is enabled disable it for testing.

Does `sysctl net.ipv4.conf.all.route_localnet`  return 1 or 0 ? If 0 set it to 1 with `systemctl -w net.ipv4.conf.all.route_localnet=1` or edit `/etc/sysctl.conf`

If you manage your VM with virsh add `trustGuestRxFilters='yes'` to your network card like this:
```
    <interface type='network' trustGuestRxFilters='yes'>
      <mac address='52:54:00:d1:24:8e'/>
      <source network='private'/>
      <model type='virtio'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x08' function='0x0'/>
    </interface>
```

If you are using virtio ensure you have a private network between your host and guest like the one above. Macvtap doesn't allow for direct host-to-guest communication unless you have a switch that support hairpinning.

**Comment by bcHelix** [comment]
I'm using Devuan (non-systemD Debian fork), and I have an NFT firewall up. I did a `nft flush ruleset` and I immediatly got sound. Very stange as NFT should not be restricting any traffic on the bridge. I'll have to take a closer look at the rules. I did make the `trustGuestRxFilters='yes'` and `sysctl -w net.ipv4.conf.all.route_localnet=1` so it could also be that they had something to do with it.

Thanks very much for your help @martinellimarco!

**Comment by coldmilk138** [comment]
Hi, I'm currently experiencing a similar issue, as detailed above I've set `trustGuestRxFilters='yes'` and done `systemctl -w net.ipv4.conf.all.route_localnet=1`, but I still don't seem to be able to get any audio output.
I tried the tcpdump step using `sudo tcpdump -pni virbr0 port 4010`, and it clearly has output there - what could the issue be?

To clarify though I also set `UnicastIPv4` to `172.16.0.1` and `UnicastPort` to `4010` to make sure (though I think the problem is on the receiving end).

In the Linux client I'm using `./scream -u -i virbr0` to hopefully get the output additionally.

**Comment by canselcik** [comment]
It's been quite a while since I looked into this however have you guys suffering with latency on the shared memory approach considered/heard of the IVSHMEM version with a bell? I remember using that for the last few years and that had all the latency issues taken care of.

**Comment by gralco** [comment]
I had this issue on Fedora 36 and fixed it. Here was my solution:

Edit the file `/usr/lib/firewalld/zones`

Remove the line that says \<reject/\> :
```
<rule priority='32767'>
  <reject/>
</rule>
```

Run `sudo systemctl restart firewalld`
